src/java.base/share/classes/jdk/internal/foreign/BufferStack.java

```diff
@@ -102,27 +119,41 @@ public Arena pushFrame(long size, long byteAlignment) {
             return new Frame(needsLock, size, byteAlignment);
         }
 
-        static PerThread of(long size) {
+        static PerThread of(long byteSize, long byteAlignment) {
             final Arena arena = Arena.ofAuto();
-            final SlicingAllocator stack = new SlicingAllocator(arena.allocate(size));
-            return new PerThread(new ReentrantLock(), stack, arena);
+            return new PerThread(new ReentrantLock(),
+                    arena,
+                    new SlicingAllocator(arena.allocate(byteSize, byteAlignment)),
+                    new CleanupAction(arena));
```

Comment:
any reason why you didn't use a lambda here?

Reply:
Also, not a big fan of records here -- it seems that many implementation details such as cleanup action, lock and slicing allocator are "leaked out" to the caller, that is now responsible to set things up correctly. I think a `PerThread` constructor with arena, size, alignment would be make the code less coupled and more readable.

Reply:
And, if you do that, you then don't need the `of` factory -- clients can just use `new`

Reply:
> any reason why you didn't use a lambda here?

I also think that CleanupAction should be changed to lambda

Reply:
Using an anonymous class for the cleanup action had very adverse effects on performance. I didn't want to use a lambda for startup performance reasons.

Reply:
If using lambda affects performance, how about using anonymous classes?
```java
            return new PerThread(new ReentrantLock(),
                    arena,
                    new SlicingAllocator(arena.allocate(byteSize, byteAlignment)),
                    new Consumer<MemorySegment>() {
                        @Override
                        public void accept(MemorySegment memorySegment) {
                            Reference.reachabilityFence(arena);
                        }});
```

Reply:
Anonymous classes also captures outer variables and break cleaner/GC. The only safe measures are local enums or records (anonymous classes cannot be enum or records)